### PR TITLE
Fix key length test and HMAC buffer size

### PR DESCRIPTION
**Pull request by bill1600**
There is no requirement for a particular key size, so we should not be enforcing it.  The key/password can be anything the user wants.  Internally, the HMAC function will pad or truncate as necessary.

Also, we need to provide a buffer for the HMAC function that can hold the maximum digest size.

### CODE REVIEW

File: libjwt/jwt.c [hunk 1]
```diff
@@ -102,7 +102,7 @@ static void jwt_scrub_key(jwt_t *jwt)
 
 int jwt_set_alg(jwt_t *jwt, jwt_alg_t alg, unsigned char *key, int len)
 {
-	int key_len = jwt_alg_key_len(alg);
+	//int key_len = jwt_alg_key_len(alg);
```
Comment (benmcollins) [inline libjwt/jwt.c]: Please just remove lines instead of commenting them out to reduce cruft.

File: libjwt/jwt.c [hunk 2]
```diff
@@ -116,23 +116,27 @@ int jwt_set_alg(jwt_t *jwt, jwt_alg_t alg, unsigned char *key, int len)
 	case JWT_ALG_HS256:
 	case JWT_ALG_HS384:
 	case JWT_ALG_HS512:
-		if (!key || len != key_len)
+		if (!key)
 			return EINVAL;
+		// key length is not required to equal jwt_alg_key_len (msg digest len)
+		// HMAC algorithm will either pad or truncate the key as necessary.
+		// if (!key || len != key_len)
+		// 	return EINVAL;
```
Comment (benmcollins) [inline libjwt/jwt.c]: Same here. Also, use /* comment */ styles, even on multi-line (see other parts of the code for examples).

File: libjwt/jwt.c [hunk 3]
```diff
@@ -302,7 +306,9 @@ static json_t *jwt_b64_decode(char *src)
 static int jwt_sign_sha_hmac(jwt_t *jwt, BIO *out, const EVP_MD *alg,
 			     const char *str)
 {
-	unsigned char res[jwt->key_len];
+  // HMAC function outputs a message digest, so we need a buffer
+  // of max digest size.
+	unsigned char res[EVP_MAX_MD_SIZE];
```
Comment (benmcollins) [inline libjwt/jwt.c]: Remove this entire comment. Not needed.

File: libjwt/jwt.c [hunk 4]
```diff
@@ -363,16 +369,19 @@ static int jwt_verify_head(jwt_t *jwt, char *head)
...
 
 		if (jwt->key) {
+#if 0
+ 		  int len;
+      // key length is not required to equal jwt_alg_key_len (msg digest len)
+      // HMAC algorithm will either pad or truncate the key as necessary.
 			len = jwt_alg_key_len(jwt->alg);
 			if (len != jwt->key_len)
 				ret = EINVAL;
+#endif
```
Comment (benmcollins) [inline libjwt/jwt.c]: Just remove all of these to keep code from getting crufty.

File: tests/jwt_encode.c [hunk 5]
```diff
@@ -223,13 +223,13 @@ START_TEST(test_jwt_encode_invalid)
...
-	ck_assert_int_eq(ret, EINVAL);
+	ck_assert_int_eq(ret, 0);
 
 	ret = jwt_set_alg(jwt, JWT_ALG_HS256, key512, 64);
-	ck_assert_int_eq(ret, EINVAL);
+	ck_assert_int_eq(ret, 0);
 
 	ret = jwt_set_alg(jwt, JWT_ALG_HS384, key512, 16);
-	ck_assert_int_eq(ret, EINVAL);
+	ck_assert_int_eq(ret, 0);
```
Comment (benmcollins) [inline tests/jwt_encode.c]: Please change the name of this test so it reflects the new expected behavior.

**Comment by benmcollins** [comment]
Thanks for the patch! Please see comments. Once fixed, I'll merge the change.

**Comment by bill1600** [comment]
OK.
Fixed comments.

**Comment by benmcollins** [comment]
This has diverged significantly from master. Please break out into separate works to make merging easier and submit new pull requests.

Thanks!

**Pull request closed without merging**